File: Vowel Spellchecker/Vowel Spellchecker.java
// Solution for Vowel Spellchecker in JAVA

class Solution {
    // Helper: Replace vowels with '*'
    private String maskVowels(String word) {
        StringBuilder sb = new StringBuilder();
        for (char ch : word.toCharArray()) {
            if ("aeiou".indexOf(ch) != -1) {
                sb.append('*');
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }

    public String[] spellchecker(String[] wordlist, String[] queries) {
        // Exact word set
        Set<String> exactMatch = new HashSet<>();

        // Case-insensitive and vowel-insensitive maps
        Map<String, String> caseInsensitiveMatch = new HashMap<>();
        Map<String, String> vowelInsensitiveMatch = new HashMap<>();

        // Preprocess wordlist
        for (String word : wordlist) {
            exactMatch.add(word);

            String lowerWord = word.toLowerCase();

            caseInsensitiveMatch.putIfAbsent(lowerWord, word);

            String vowelMasked = maskVowels(lowerWord);
            vowelInsensitiveMatch.putIfAbsent(vowelMasked, word);
        }

        String[] result = new String[queries.length];

        // Process each query
        for (int i = 0; i < queries.length; i++) {
            String query = queries[i];

            if (exactMatch.contains(query)) {
                result[i] = query;
                continue;
            }

            String lowerQuery = query.toLowerCase();
            if (caseInsensitiveMatch.containsKey(lowerQuery)) {
                result[i] = caseInsensitiveMatch.get(lowerQuery);
                continue;
            }

            String vowelMaskedQuery = maskVowels(lowerQuery);
            if (vowelInsensitiveMatch.containsKey(vowelMaskedQuery)) {
                result[i] = vowelInsensitiveMatch.get(vowelMaskedQuery);
                continue;
            }

            result[i] = "";
        }

        return result;
    }
}

File: Vowel Spellchecker/Vowel Spellchecker.py
# Solution for Vowel Spellchecker in PY

class Solution:
    def maskVowels(self, word: str) -> str:
        return "".join('*' if ch in "aeiou" else ch for ch in word)

    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        # Exact match set
        exactMatch = set(wordlist)

        # Case-insensitive and vowel-insensitive maps
        caseInsensitiveMatch = {}
        vowelInsensitiveMatch = {}

        # Preprocess wordlist
        for word in wordlist:
            lowerWord = word.lower()

            if lowerWord not in caseInsensitiveMatch:
                caseInsensitiveMatch[lowerWord] = word

            vowelMasked = self.maskVowels(lowerWord)
            if vowelMasked not in vowelInsensitiveMatch:
                vowelInsensitiveMatch[vowelMasked] = word

        result = []

        # Process queries
        for query in queries:
            if query in exactMatch:
                result.append(query)
                continue

            lowerQuery = query.lower()
            if lowerQuery in caseInsensitiveMatch:
                result.append(caseInsensitiveMatch[lowerQuery])
                continue

            vowelMaskedQuery = self.maskVowels(lowerQuery)
            if vowelMaskedQuery in vowelInsensitiveMatch:
                result.append(vowelInsensitiveMatch[vowelMaskedQuery])
                continue

            result.append("")

        return result
